tracer/src/Datadog.Trace/RuntimeMetrics/MemoryMappedCounters.cs

```diff
@@ -0,0 +1,593 @@
+    {
+        private const string GarbageCollectionMetrics = $"{MetricsNames.Gen0HeapSize}, {MetricsNames.Gen1HeapSize}, {MetricsNames.Gen2HeapSize}, {MetricsNames.LohSize}, {MetricsNames.ContentionCount}, {MetricsNames.Gen0CollectionsCount}, {MetricsNames.Gen1CollectionsCount}, {MetricsNames.Gen2CollectionsCount}";
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<PerformanceCountersListener>();
+
+        private readonly IDogStatsd _statsd;
+        private readonly int _processId;
+
+        private int? _previousGen0Count;
+        private int? _previousGen1Count;
+        private int? _previousGen2Count;
+
+        private double? _lastContentionCount;
+        private MemoryMappedFile _file;
+        private MemoryMappedViewAccessor _view;
+
+        public MemoryMappedCounters(IDogStatsd statsd)
+        {
+            _statsd = statsd;
+
+            ProcessHelpers.GetCurrentProcessInformation(out _, out _, out _processId);
+
+            MemoryMappedFile file = null;
+            MemoryMappedViewAccessor view = null;
+
+            try
+            {
+                file = MemoryMappedFile.OpenExisting(@"Cor_CLR_WRITER\Cor_SxSPublic_IPCBlock");
+                view = file.CreateViewAccessor();
+
+                if (GC.CollectionCount(0) == 0)
+                {
+                    // This is unlikely to happen, but we must make sure that the GC has run at least once,
+                    // otherwise the GC structure that we use for the sanity check will be empty.
+                    GC.Collect(0, GCCollectionMode.Forced, blocking: true); // Sorry ._.
+                }
+
+                // Sanity check
+                view.Read<IpcControlBlock>(0, out var controlBlock);
+
+                if (controlBlock.Header.Flags == 0)
+                {
+                    throw new InvalidOperationException("The IPC control block is not initialized");
+                }
+
+                if (controlBlock.Perf.GC.ProcessID != _processId)
+                {
+                    throw new InvalidOperationException($"The PID in the IPC control block does not match (expected {_processId}, found {controlBlock.Perf.GC.ProcessID}");
+                }
+            }
+            catch
+            {
+                view?.Dispose();
+                file?.Dispose();
+                throw;
+            }
+
+            _file = file;
+            _view = view;
+        }
+
+        [Flags]
+        private enum IpcHeaderFlags : ushort
+        {
+            IPC_FLAG_USES_FLAGS = 0x1,
+            IPC_FLAG_INITIALIZED = 0x2,
+            IPC_FLAG_X86 = 0x4
+        }
+
+        public void Dispose()
+        {
+            if (_view != null)
+            {
+                _view.Dispose();
+                _view = null;
+            }
+
+            if (_file != null)
+            {
+                _file.Dispose();
+                _file = null;
+            }
```

Comment:
tiny nit: I think you could end up disposing twice here technically?

Reply:
That's the good point, the documentation is not clear on whether the view gains ownership on the memory-mapped file 🤔  I'd rather keep the double dispose just in case (it looks like the MemoryMappedFile is resilient against it anyway)

Reply:
I actually meant that if `Dispose()` is called twice on `MemoryMappedCounters` you could get a double dispose, but yeah, that too 😄 

I'm not sure if that can actually happen of course... we only dispose the runtime metrics when creating a new `TracerManager`, so probably not.

Somewhat related: should we be disposing this on shutdown? does it matter? Is it better to be safe? 🤔

Reply:
> I actually meant that if Dispose() is called twice on MemoryMappedCounters you could get a double dispose

Only if the two calls to dispose are simultaneous. I don't think we should worry about that (plus, the current implementation has the same weakness). 

> Somewhat related: should we be disposing this on shutdown? does it matter? Is it better to be safe? 🤔

Honestly, we have already so many shutdown issues that I'd rather leave it as is 😄 Strictly speaking we should probably dispose the TracerManager when the appdomain is unloaded, but for the MemoryMappedCounters in particular I believe the finalizers will take care of the handle.

Reply:
Added the dispose in https://github.com/DataDog/dd-trace-dotnet/pull/3403/commits/e12b216d0269e808a4b02a95dc9791ae418e0899